homeassistant/components/vacuum/xiaomi_miio.py

```diff
@@ -234,7 +235,12 @@ def device_state_attributes(self):
                     / 3600),
                 ATTR_FILTER_LEFT: int(
                     self.consumable_state.filter_left.total_seconds()
-                    / 3600)})
+                    / 3600),
+                ATTR_SENSOR_DIRTY_LEFT: int(
+                    self.consumable_state.sensor_dirty_left.total_seconds()
```

Comment:
We never store timedelta's in the state machine. What does it contain and when does it get updated?

Reply:
`consumable_state` is an object exposed by the underlying library, which is updated on every call to `update` (we are dealing with a polling component here). It stores the remaining time of each consumable as a timedelta. This is the time until the next maintanance should take place and is decreasing as the vacuum cleaner is doing it's job.

IMHO it actually makes sense to store these kind of values as timedeltas from a Python-perspective. That's what the library does. But as you have mentioned it does not really make sense to expose these deltas to the state machine. This is why we convert them back into hours in `device_state_attributes`.

Reply:
All these values need to be changed to work like the time left for media players: the moment a vacuum is running, we store time left and we store time left updated at. As long as the state is vacuuming, the frontend can just calculate time left - time since last state change. That way we are not spamming the state machine with states.

Reply:
Ahh, now I see what you are referring to. Generally I have implemented the new attribute the same way as all of the other attributes have been implemented for the device. That's why I think your comment does not apply solely to my PR but actually to the entire component.

Since I am not the actual author/maintainer of the component I can only add some general thoughts about this (sorry if I get anything wrong here): First of all the actual values exposed to the state machine change **at most** once per hour since the unit is "hours left". I would not consider one attribute change per hour (at most) as spamming. This makes these attributes quite different from the current playback position that media players provide, which of course should be as accurate as possible (people need seconds there). So yes, right now these numbers can be 20 seconds off (the poll interval for the `vacuum` platform), which is not that bad for a unit representing "hours left".

Second of all these values are reported by the device, calculated by a proprietary algorithm. The assumption, that for every second that the vacuum cleaner is running the lifespan of it's consumables is also reduced by one second might not be correct. AFAIK the values also depend on the current fan/brush speed, eco mode, carpet mode and probably other things. The official Mi Home app also says "approximately *n* hours left", so even the vendor considers these numbers to be a rough estimate. To be safe here I think it's better to read these values directly from the device (which is done right now) because calculating them can be quite hard and error-prone for pretty much zero benefit. But it's up the maintainers to decide this.

Reply:
If it's hours left, I am fine with it as I agree that 24 changes max is not a lot.